contracts/hydro/src/slashing.rs

```diff
@@ -0,0 +1,767 @@
+                // It should never happen that a lock that voted on a proposal in the given round
+                // couldn't be found, but if it does happen the only thing we can do is skip it.
+                context.skipped_lockups.push(voted_lock.0);
+                continue;
+            }
+            Some(lockup) => lockup,
+        };
+
+        for (lock_id, fraction) in get_current_lock_composition(&deps.as_ref(), voted_lock.0)? {
+            let mut lockup_to_slash = match LOCKS_MAP_V2.may_load(deps.storage, lock_id)? {
+                None => {
+                    // Lock returned by `get_current_lock_composition()` should always exist
+                    context.skipped_lockups.push(lock_id);
+                    continue;
+                }
+                Some(lockup) => lockup,
+            };
+
+            let (amount_to_slash, _) = into_amount_to_slash(
+                &deps.as_ref(),
+                &mut token_manager,
+                &voted_lockup,
+                &lockup_to_slash,
+                fraction,
+                slash_percent,
+                round_id,
+                current_round_id,
+            )?;
+
+            // This can happen if user did vote on the given proposal, but that vote power dropped to 0 afterwards,
+            // or if we cannot obtain the ratio for the token held by the given lockup.
+            if amount_to_slash.is_zero() {
+                context.skipped_lockups.push(lock_id);
+                continue;
+            }
+
+            // Caculate the amount to be slashed by adding newly computed value to already known pending slashes.
+            // If the amount to be slashed ends up being greater than total amount held by the lockup (highly unlikely),
+            // then use the total lockup amount.
+            let amount_to_slash = LOCKS_PENDING_SLASHES
+                .may_load(deps.storage, lockup_to_slash.lock_id)?
+                .unwrap_or_default()
+                .checked_add(amount_to_slash)?
+                .min(lockup_to_slash.funds.amount);
+
+            // If the percentage of the lockup to be slashed is greater than or equalt to the
+            // slashing percentage threshold, perform the actual slashing. Otherwise, just store
+            // information about the pending slash and proceed to the next lockup to be checked.
+            if Decimal::from_ratio(amount_to_slash, lockup_to_slash.funds.amount)
+                .ge(&constants.slash_percentage_threshold)
+            {
+                // Subtract the amount to slash from the lockup.
+                lockup_to_slash.funds.amount =
+                    lockup_to_slash.funds.amount.checked_sub(amount_to_slash)?;
+
+                // If the amount left on the lockup is 0, the lockup should be removed from the store.
+                if lockup_to_slash.funds.amount == Uint128::zero() {
+                    LOCKS_MAP_V2.remove(deps.storage, lockup_to_slash.lock_id, env.block.height)?;
```

Comment:
After [this PR](https://github.com/informalsystems/hydro/pull/298) gets merged, make sure to also remove the lock_id from `TOKEN_IDS` if needed.